tests/test_main.py

```diff
@@ -0,0 +1,255 @@
+    groups = [('test-group', 'title', [create_command('cmd-name', 'cmd help')])]
+    dispatcher = Dispatcher([], groups)
+    fake_stdout = io.StringIO()
+    with patch('sys.stdout', fake_stdout):
+        retcode = dispatcher.run()
+    assert retcode == -1
+    assert "usage: charmcraft" in fake_stdout.getvalue()
+
+
+def test_dispatcher_command_execution_ok():
+    """Command execution depends of the indicated name in command line, return code ok."""
+    class MyCommandControl(BaseCommand):
+        help_msg = "some help"
+
+        def __init__(self, group):
+            super().__init__(group)
+            self.executed = None
+
+        def run(self, parsed_args):
+            self.executed = parsed_args
+
+    class MyCommand1(MyCommandControl):
+        name = 'name1'
+
+    class MyCommand2(MyCommandControl):
+        name = 'name2'
+
+    groups = [('test-group', 'title', [MyCommand1, MyCommand2])]
+    dispatcher = Dispatcher(['name2'], groups)
+    retcode = dispatcher.run()
+    assert dispatcher.commands['name1'].executed is None
+    assert isinstance(dispatcher.commands['name2'].executed, argparse.Namespace)
+    assert retcode == 0
+
+
+def test_dispatcher_command_execution_crash():
+    """Command crashing is let pass through."""
+    class MyCommand(BaseCommand):
+        help_msg = "some help"
+        name = 'cmdname'
+
+        def run(self, parsed_args):
+            raise ValueError()
+
+    groups = [('test-group', 'title', [MyCommand])]
+    dispatcher = Dispatcher(['cmdname'], groups)
+    with pytest.raises(ValueError):
+        dispatcher.run()
+
+
+def test_dispatcher_command_execution_controlled_error():
+    """Commands can indicate "fatal error" through a specific exception."""
+    class MyCommand(BaseCommand):
+        help_msg = "some help"
+        name = 'cmdname'
+
+        def run(self, parsed_args):
+            raise CommandError("boom", retcode=-13)
+
+    groups = [('test-group', 'title', [MyCommand])]
+    dispatcher = Dispatcher(['cmdname'], groups)
+    retcode = dispatcher.run()
+    assert retcode == -13
+
+
+@pytest.mark.parametrize("option", ['--verbose', '-v'])
+def test_dispatcher_generic_setup_verbose(option):
+    """Generic parameter handling for verbose log setup."""
+    dispatcher = Dispatcher([option], [])
+    with patch('charmcraft.logsetup.configure') as mock:
+        dispatcher.run()
+    mock.assert_called_with('verbose')
+
+
+@pytest.mark.parametrize("option", ['--quiet', '-q'])
+def test_dispatcher_generic_setup_quiet(option):
+    """Generic parameter handling for silent log setup."""
+    dispatcher = Dispatcher([option], [])
+    with patch('charmcraft.logsetup.configure') as mock:
+        dispatcher.run()
+    mock.assert_called_with('quiet')
+
+
+def test_dispatcher_load_commands_ok():
+    """Correct command loading."""
+    cmd1, cmd2, cmd3 = [create_command('cmd-name-{}'.format(n), 'cmd help') for n in range(3)]
```

Comment:
its a little odd that n starts at 0 but cmd1 starts at 1.
cmd0, cmd1, cmd2 

would at least line them up?

Reply:
Done.